File: lab2/gr9sub1/TOPCHEEVA_TAISIA/src/fork_example.py
import os
import sys

def main():
    print("Parent starting", flush=True)

    pid1 = os.fork()
    if pid1 == -1:
        print("fork failed", file=sys.stderr)
        sys.exit(1)
    if pid1 == 0:
        print(f"child[0]: PID={os.getpid()}, PPID={os.getppid()}", flush=True)
        sys.exit(0)

    pid2 = os.fork()
    if pid2 == -1:
        print("fork failed", file=sys.stderr)
        sys.exit(1)
    if pid2 == 0:
        print(f"child[1]: PID={os.getpid()}, PPID={os.getppid()}", flush=True)
        sys.exit(0)

    pid, status = os.waitpid(pid1, 0)
    print(f"Child {pid} exited with code {os.WEXITSTATUS(status)}", flush=True)
    pid, status = os.waitpid(pid2, 0)
    print(f"Child {pid} exited with code {os.WEXITSTATUS(status)}", flush=True)
    print("Parent finished", flush=True)

if __name__ == "__main__":
    main()

File: lab2/gr8sub1/Стёпкин_Владимир/src/fork_example.c
#include <stdio.h>
#include <unistd.h>
#include <sys/wait.h>
#include <stdlib.h>

int main() {
    printf("parent(PID=%d) started\n", getpid());
    fflush(stdout);

    pid_t child_pids[2];

    for (int i = 0; i < 2; i++) {
        pid_t pid = fork();
        if (pid < 0) {
            perror("fork failed");
            exit(1);
        } else if (pid == 0) {
            // Дочерний процесс
            printf("child_%d: PID=%d, PPID=%d\n", i, getpid(), getppid());
            fflush(stdout);
            exit(0);  // дочерний завершается
        } else {
            // Родительский процесс
            child_pids[i] = pid;
        }
    }

    // Родитель ждёт дочерних процессов
    for (int i = 0; i < 2; i++) {
        int status;
        pid_t wpid = waitpid(child_pids[i], &status, 0);
        if (wpid > 0) {
            if (WIFEXITED(status)) {
                printf("child with PID %d exited with code %d\n", wpid, WEXITSTATUS(status));
            } else {
                printf("child with PID %d terminated abnormally\n", wpid);
            }
            fflush(stdout);
        }
    }

    printf("parent(PID=%d) finished\n", getpid());
    fflush(stdout);

    return 0;
}
